### DISCUSSION TITLE
Declaring ref locals in for loop initializer

### DISCUSSION THREAD

**Discussion by jnm2** [dotnet/csharplang#924] [General]
The ref returns and locals spec is not linked from https://github.com/dotnet/csharplang/blob/master/Language-Version-History.md#c-70-visual-studio-2017 and I'm failing to find it by searching. Because I can't find that trail, I have no idea if this is a limitation of the spec; it almost seems like a bug in Roslyn.

I expect to be able to write this:
```cs
for (ref var en = ref enumerator; en.MoveNext();)
{
   // en.Current
}
```

Since I can write this:
(I thought this didn't work because I mistakenly had `enumerator` as a readonly class field, but the point still remains.)

```cs
for (var en = enumerator; en.MoveNext();)
{
   // en.Current
}
```

The workaround is to stop scoping `en` to the loop:

```cs
ref var en = ref enumerator;
while (en.MoveNext())
{
   // en.Current
}
```

You may dispute the liberties I tend to take with `for` loops, but I could construct a similar example that uses iteration, condition and iterator.

The point is that it seems wrong that a valid local declaration statement cannot be used in the `
for_initializer`:

```antlr
for_initializer
    : local_variable_declaration
    | statement_expression_list
    ;
```

https://github.com/dotnet/csharplang/blob/master/spec/statements.md#the-for-statement

**Comment by HaloFour** [comment]
@jnm2 

> Since I can write this (though this doesn't work because the enumerator is a struct which mutates on MoveNext):

Why do you need a `ref` for that?  Structs have always been allowed to mutate themselves via instance methods:

```cs
Point pt = new Point(2, 4);
pt.Offset(2, 2); // self-mutating method
Debug.Assert(pt.X == 4 && pt.Y == 6);
```

Any time you call a struct instance method the `this` argument is actually a `ref`.

**Comment by jnm2** [comment]
@HaloFour That's what I thought, but I observed this not being the case when I moved the struct enumerator to a class field. I'll create a code example.

**Comment by HaloFour** [comment]
@jnm2 

Was the field `readonly`?  If it is the C# compiler will try to protect you by assigning the field to a local first and passing the `ref` to the local to the instance method.  That way if the method mutates the local would be affected and not the field.

**Comment by jnm2** [comment]
@HaloFour Yes, that must be what happened. Just tried it both ways. So I don't actually need ref locals for anything yet... I was so happy there for a moment 😜 

But my dislike of the inconsistency in `for_initializer` still stands.
